### PR TITLE
Init

### PR DESCRIPTION
**Pull request by delvedor**
Feedbacks?

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+language: node_js
+
+node_js:
+  - "10"
+  - "9"
```
Comment (mcollina) [inline .travis.yml]: don't test on 9

File: index.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    } else {
+      var result = validate(credentials.name, credentials.pass, req, res, done)
+      if (result && typeof result.then === 'function') {
+        result.then(done, done)
+      }
+    }
+
+    function done (err) {
+      if (err !== undefined) {
+        unauthorized(err.message)
```
Comment (mcollina) [inline index.js]: I'm not 100% happy about this. I think we should have two different paths, one for errors and one for validation. Or we respect the fact that if `err.statusCode !== 401`  we use that code and error message.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        next()
+      }
+    }
+
+    function unauthorized (message) {
+      var payload = serializeError({
+        error: 'Unauthorized',
+        message: message || 'Access denied',
+        statusCode: 401
+      })
```
Comment (mcollina) [inline index.js]: I would prefer if we used `res.send(err)`.

File: index.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+function basicPlugin (fastify, opts, next) {
+  if (typeof opts.validate !== 'function') {
+    return next(new Error('Basic Auth: Missing validate function'))
+  }
+
+  const validate = opts.validate
+  fastify.decorate('basicAuth', basicAuth)
+  if (opts.disableHook !== true) {
+    fastify.addHook('onRequest', onBasicAuth)
```
Comment (mcollina) [inline index.js]: Can this be a `preHandler`  instead?

File: test.js [hunk 5]
```diff
@@ -0,0 +1,338 @@
+'use strict'
+
+const { test } = require('tap')
+const Fastify = require('fastify')
+const basicAuth = require('./index')
```
Comment (mcollina) [inline test.js]: Can you add a test where you demo this with https://github.com/fastify/fastify-auth?

File: index.js (hunk 3 above)
Comment (delvedor) [reply index.js]: Nope, we are inside a `onRequest` hook and we can't access the `reply` interface.

File: index.js (hunk 4 above)
Comment (delvedor) [reply index.js]: I don't know which one is better.

File: README.md [hunk 6]
```diff
@@ -1 +1,97 @@
...
+npm i fastify-basic-auth
+```
+## Usage
+This plugin decorates the fastifyinstance with a `basicAuth` function, which you can use inside a `preHandler` hook, in a `beforeHander` or with [`fastify-auth`](https://github.com/fastify/fastify-auth).
+
+```js
+const fastify = require('fastify')()
+
+fastify.register(require('fastify-basic-auth'), { validate })
+function validate (username, password, req, res, done) {
```
Comment (mcollina) [inline README.md]: are these req and res, or request and reply from fastify?

File: test.js [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+  function validate (username, password, req, res, done) {
+    if (username === 'user' && password === 'pwd') {
+      done()
+    } else {
+      done(new Error('Winter is coming'))
+    }
+  }
+
+  fastify.after(() => {
+    fastify.addHook('preHandler', fastify.auth([fastify.basicAuth]))
```
Comment (mcollina) [inline test.js]: Can you also add this in the README?

**Pull request merged**